sdk/data/azcosmos/cosmos_global_endpoint_manager.go

```diff
@@ -0,0 +1,170 @@
+	if err != nil {
+		return accountProperties{}, fmt.Errorf("failed to send GET request: %w", err)
+	}
+
+	properties, err := parseAccountProperties(azResponse)
+	if err != nil {
+		return accountProperties{}, fmt.Errorf("failed to parse account properties: %w", err)
+	}
+
+	return properties, nil
+}
+
+// parseAccountProperties parses the account properties from the HTTP response.
+func parseAccountProperties(azResponse *http.Response) (accountProperties, error) {
+	properties := accountProperties{}
+	err := azruntime.UnmarshalAsJSON(azResponse, &properties)
+	if err != nil {
+		return properties, err
+	}
+	return properties, nil
+}
+
+// GetLocation returns the location for the given endpoint from the location cache.
+func (gem *globalEndpointManager) GetLocation(endpoint url.URL) string {
+	return gem.locationCache.getLocation(endpoint)
+}
+
+// MarkEndpointUnavailableForRead marks an endpoint as unavailable for read operations.
+func (gem *globalEndpointManager) MarkEndpointUnavailableForRead(endpoint url.URL) error {
+	return gem.locationCache.markEndpointUnavailableForRead(endpoint)
+}
+
+// MarkEndpointUnavailableForWrite marks an endpoint as unavailable for write operations.
+func (gem *globalEndpointManager) MarkEndpointUnavailableForWrite(endpoint url.URL) error {
+	return gem.locationCache.markEndpointUnavailableForWrite(endpoint)
+}
+
+// Update updates the location cache and the client's default endpoint based on the provided account properties.
+func (gem *globalEndpointManager) Update(accountProperties accountProperties) error {
+	if err := gem.locationCache.update(accountProperties.WriteRegions, accountProperties.ReadRegions, gem.preferredRegions, &accountProperties.EnableMultipleWriteLocations); err != nil {
+		return fmt.Errorf("failed to update location cache: %w", err)
+	}
+
+	gem.stopBackgroundRefresh()
+	gem.startBackgroundRefresh(accountProperties)
+
+	endpoints, err := gem.locationCache.writeEndpoints()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve write endpoints: %w", err)
+	}
+	if len(endpoints) > 0 {
+		gem.client.endpoint = endpoints[0].String()
+	}
+
+	return nil
+}
+
+func (gem *globalEndpointManager) startBackgroundRefresh(accountProperties accountProperties) {
+	gem.Lock()
+	defer gem.Unlock()
+
+	gem.refreshTicker = time.NewTicker(DefaultBackgroundRefreshInterval)
+	gem.refreshDone = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-gem.refreshTicker.C:
+				err := gem.locationCache.update(accountProperties.WriteRegions, accountProperties.ReadRegions, gem.preferredRegions, &accountProperties.EnableMultipleWriteLocations)
+				if err != nil {
+					log.Write(azlog.EventResponse, fmt.Sprintf("Background refresh error: %v", err))
+				}
+			case <-gem.refreshDone:
+				gem.refreshTicker.Stop()
+				close(gem.refreshDone)
+				return
+			}
+		}
+	}()
+}
+
+func (gem *globalEndpointManager) stopBackgroundRefresh() {
```

Comment:
What is the use case for stopping the background refresh?

Reply:
The stopBackgroundRefresh method is called when you explicitly invoke it on an instance of the globalEndpointManager. It is typically used to stop the background refresh of the location cache when it is no longer needed or when you want to manually control the refresh process.

The stopBackgroundRefresh method is called within the newGlobalEndpointManager function, right after creating a new instance of the globalEndpointManager. It ensures that any existing background refresh process is stopped before starting a new one.

Additionally, the stopBackgroundRefresh method is designed to be called at any point in your application's lifecycle when you want to stop the background refresh process. For example, you might call it when shutting down your application or when specific conditions are met.

By calling stopBackgroundRefresh, you stop the periodic refresh of the location cache performed by the background goroutine, as specified by the refreshTicker. This helps conserve resources and prevent unnecessary updates to the location cache when they are not required.